src/axom/spin/internal/linear_bvh/TraversalPredicates.hpp

```diff
@@ -0,0 +1,184 @@
+   *
+   */
+  template < typename PointType >
+  AXOM_HOST_DEVICE
+  static inline bool pointInLeftBin( const PointType& point,
+                                     const VecType< FloatType >& s1,
+                                     const VecType< FloatType >& s2 ) noexcept;
+
+  /*!
+   * \brief Checks if the supplied point is within the right bin.
+   *
+   * \param [in] point the coordinates of the point in query.
+   * \param [in] s2 the 2nd segment of the BVH that stores the right bin.
+   * \param [in] s3 the 3rd segment of the BVH that stores the right bin.
+   *
+   * \return status true if the point is inside the right bin, else, false.
+   */
+  template < typename PointType >
+  AXOM_HOST_DEVICE
+  static inline bool pointInRightBin( const PointType& point,
+                                      const VecType< FloatType >& s2,
+                                      const VecType< FloatType >& s3 ) noexcept;
+
+  /// @}
+
+};
+
+constexpr int DIMENSION_2 = 2;
+constexpr int DIMENSION_3 = 3;
+
+//------------------------------------------------------------------------------
+// 2D Specialization
+//------------------------------------------------------------------------------
+template < typename FloatType >
+class TraversalPredicates< DIMENSION_2, FloatType >
+{
+
+public:
+
+  template < typename PointType >
+  AXOM_HOST_DEVICE
+  static inline bool pointInLeftBin( const PointType& point,
+                                     const VecType< FloatType >& s1,
+                                     const VecType< FloatType >& s2  ) noexcept
+  {
+    bool in_left = true;
+    if ( point[0] < s1[0] ) in_left = false;
+    if ( point[1] < s1[1] ) in_left = false;
+
+    if ( point[0] > s1[3] ) in_left = false;
+    if ( point[1] > s2[0] ) in_left = false;
+
+    return in_left;
+  }
+
+  template < typename PointType >
+  AXOM_HOST_DEVICE
+  static inline bool pointInRightBin( const PointType& point,
+                                      const VecType< FloatType >& s2,
+                                      const VecType< FloatType >& s3  ) noexcept
+  {
+    bool in_right = true;
+
+    if ( point[0] < s2[2] ) in_right = false;
+    if ( point[1] < s2[3] ) in_right = false;
+
+    if ( point[0] > s3[1] ) in_right = false;
+    if ( point[1] > s3[2] ) in_right = false;
+
+    return in_right;
+  }
+
+};
+
+//------------------------------------------------------------------------------
+// 3D Specialization
+//------------------------------------------------------------------------------
+template < typename FloatType >
+class TraversalPredicates< DIMENSION_3, FloatType >
+{
+
+public:
+
+  template < typename PointType >
+  AXOM_HOST_DEVICE
+  static inline bool pointInLeftBin( const PointType& point,
+                                     const VecType< FloatType >& s1,
+                                     const VecType< FloatType >& s2  ) noexcept
+  {
+
+    bool in_left = true;
```

Comment:
@gzagaris please confirm that the point and s* indexing here and in the routine below are correct. It looks a little odd to me. It may very well be correct, but I have no way to assess. Would some brief comments be helpful for others who may have to work in this code in the future?

Reply:
Yes. This is correct. It is due to the underlying internal layout of the BVH. I agree it looks odd, but after discussing with @mclarsen a while ago, I have a better appreciation of the reasons behind this design decision and the potential for performance gains with this layout.

It is documented here: 
https://github.com/LLNL/axom/blob/146f2b12fa1ffe4d430a449e13adece30d1a5de2/src/axom/spin/internal/linear_bvh/BVHData.hpp#L26-L54

Reply:
OK. I missed this. Maybe put a one-line comment in the implementation referencing this?

Reply:
Certainly, I'll put a comment

Reply:
Done.

Reply:
Thanks!